### PR TITLE
CSHARP-2032: Drop support for MongoDB 2.4

### PR DESCRIPTION
**Pull request by MikalaiMazurenka**
Evergreen: https://evergreen.mongodb.com/version/5e32c588306615182aac9098

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs [hunk 1]
```diff
@@ -155,7 +155,9 @@ internal IWriteOperation<BsonDocument> CreateOperation(SemanticVersion serverVer
             }
             else
             {
+#pragma warning disable 618
                 return new CreateIndexesUsingInsertOperation(_collectionNamespace, _requests, _messageEncoderSettings);
```
Comment (MikalaiMazurenka) [inline src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs]: Cannot remove this option because the alternative option (line 148) `Feature.CreateIndexesCommand` starts support from 3.0 server version.

File: src/MongoDB.Driver.Core/Core/Operations/AggregateOperation.cs [hunk 2]
```diff
@@ -253,6 +253,7 @@ public bool RetryRequested
         /// <value>
         /// A value indicating whether the server should use a cursor to return the results.
         /// </value>
+        [Obsolete("This property is no longer used and will be removed in a later release.")]
```
Comment (jyemin) [inline src/MongoDB.Driver.Core/Core/Operations/AggregateOperation.cs]: It is really no longer used, or no longer used since some server version?

File: src/MongoDB.Driver.Core/Core/Operations/AggregateOperation.cs [hunk 3]
```diff
@@ -343,16 +344,13 @@ internal BsonDocument CreateCommand(ConnectionDescription connectionDescription,
                 { "readConcern", readConcern, readConcern != null }
             };
 
-            if (Feature.AggregateCursorResult.IsSupported(connectionDescription.ServerVersion))
+            var useCursor = _useCursor.GetValueOrDefault(true) || connectionDescription.ServerVersion >= new SemanticVersion(3, 5, 0);
```
Comment (jyemin) [inline src/MongoDB.Driver.Core/Core/Operations/AggregateOperation.cs:346]: Use released version, e.g. `3, 6, 0`

File: src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs (hunk 1 above)
Comment (jyemin) [reply src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs]: Why is the pragma new?

File: src/MongoDB.Driver.Core/Core/Operations/DeleteOpcodeOperation.cs [hunk 4]
```diff
@@ -26,6 +26,7 @@ namespace MongoDB.Driver.Core.Operations
     /// <summary>
     /// Represents a delete operation using the delete opcode.
     /// </summary>
+    [Obsolete("Use BulkMixedWriteOperation instead.")]
```
Comment (jyemin) [inline src/MongoDB.Driver.Core/Core/Operations/DeleteOpcodeOperation.cs]: Let's consider whether we can just delete this class (and the others like it), despite the fact that it's public.

File: src/MongoDB.Driver.Core/Core/WireProtocol/DeleteWireProtocol.cs [hunk 5]
```diff
@@ -13,15 +13,16 @@
...
 using MongoDB.Bson;
 using MongoDB.Driver.Core.Connections;
 using MongoDB.Driver.Core.Misc;
-using MongoDB.Driver.Core.Operations;
 using MongoDB.Driver.Core.WireProtocol.Messages;
 using MongoDB.Driver.Core.WireProtocol.Messages.Encoders;
 
 namespace MongoDB.Driver.Core.WireProtocol
 {
+    [Obsolete("Use CommandWireProtocol instead.")]
```
Comment (jyemin) [inline src/MongoDB.Driver.Core/Core/WireProtocol/DeleteWireProtocol.cs]: Consider deleting this class

File: src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/BinaryEncoders/BinaryMessageEncoderFactory.cs [hunk 6]
```diff
@@ -77,7 +77,9 @@ public IMessageEncoder GetCompressedMessageEncoder(IMessageEncoderSelector origi
         /// <inheritdoc/>
         public IMessageEncoder GetDeleteMessageEncoder()
```
Comment (jyemin) [inline src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/BinaryEncoders/BinaryMessageEncoderFactory.cs]: When would this method be called now?

**Review by jyemin — COMMENTED** [review]
Before going any further, let's get consensus on whether some of the classes/methods in Core can just be deleted instead of Obsoleted.

File: src/MongoDB.Driver.Core/Core/Operations/DeleteOpcodeOperation.cs (hunk 4 above)
Comment (MikalaiMazurenka) [reply src/MongoDB.Driver.Core/Core/Operations/DeleteOpcodeOperation.cs]: I would say we shouldn't delete these classes, because of the following:

- `DeleteOpcodeOperation` is used in `MongoCollection.Remove` public method (https://github.com/mongodb/mongo-csharp-driver/blob/2ad99632b9af100060ce14288950114718b99e0c/src/MongoDB.Driver.Legacy/MongoCollection.cs#L1871)
- Similar for `UpdateOpcodeOperation`.
- `InsertOpcodeOperation` is used in `MongoCollection.Insert` (https://github.com/mongodb/mongo-csharp-driver/blob/2ad99632b9af100060ce14288950114718b99e0c/src/MongoDB.Driver.Legacy/MongoCollection.cs#L2452) which is in turn is used in public class `MongoGridFSStream` (https://github.com/mongodb/mongo-csharp-driver/blob/2ad99632b9af100060ce14288950114718b99e0c/src/MongoDB.Driver.Legacy/GridFS/MongoGridFSStream.cs#L483) and in a lot of Legacy tests (e.g. https://github.com/mongodb/mongo-csharp-driver/blob/2ad99632b9af100060ce14288950114718b99e0c/tests/MongoDB.Driver.Legacy.Tests/Builders/QueryBuilderTests.cs#L888). Additionally `InsertOpcodeOperation` is used in `CreateIndexesUsingInsertOperation` (https://github.com/mongodb/mongo-csharp-driver/blob/2ad99632b9af100060ce14288950114718b99e0c/src/MongoDB.Driver.Core/Core/Operations/CreateIndexesUsingInsertOperation.cs#L130) which is not from Legacy.

File: src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs (hunk 1 above)
Comment (MikalaiMazurenka) [reply src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs]: If I correctly understand the question, `#pragma` was added because `CreateIndexesUsingInsertOperation` was deprecated.

File: src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/BinaryEncoders/BinaryMessageEncoderFactory.cs (hunk 6 above)
Comment (MikalaiMazurenka) [reply src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/BinaryEncoders/BinaryMessageEncoderFactory.cs]: My bad, missed this factory. I will deprecate this method along with update and insert.

File: src/MongoDB.Driver.Core/Core/WireProtocol/DeleteWireProtocol.cs (hunk 5 above)
Comment (MikalaiMazurenka) [reply src/MongoDB.Driver.Core/Core/WireProtocol/DeleteWireProtocol.cs]: Same as for `DeleteOpcodeOperation` it is used in `Server.Delete` public method (https://github.com/mongodb/mongo-csharp-driver/blob/2ad99632b9af100060ce14288950114718b99e0c/src/MongoDB.Driver.Core/Core/Servers/Server.cs#L661) which is used in `DeleteOpcodeOperation` (https://github.com/mongodb/mongo-csharp-driver/blob/2ad99632b9af100060ce14288950114718b99e0c/src/MongoDB.Driver.Core/Core/Operations/DeleteOpcodeOperation.cs#L190).
Thus, if we do not delete `DeleteOpcodeOperation`, `DeleteWireProtocol` shouldn't be deleted as well.

File: src/MongoDB.Driver.Core/Core/Operations/AggregateOperation.cs (hunk 3 above)
Comment (MikalaiMazurenka) [reply src/MongoDB.Driver.Core/Core/Operations/AggregateOperation.cs:346]: I will update this to use 3.6.0.

File: src/MongoDB.Driver.Core/Core/Operations/AggregateOperation.cs (hunk 2 above)
Comment (MikalaiMazurenka) [reply src/MongoDB.Driver.Core/Core/Operations/AggregateOperation.cs]: According to documentation (https://docs.mongodb.com/v3.6/reference/command/aggregate/)
`UseCursor` was deprecated in server version 3.4 and removed in 3.6 (thus making `cursor` mandatory). I will update the message to reflect that.

File: src/MongoDB.Driver.Core/Core/Operations/DeleteOpcodeOperation.cs (hunk 4 above)
Comment (jyemin) [reply src/MongoDB.Driver.Core/Core/Operations/DeleteOpcodeOperation.cs]: @rstam do you now why that is?  Is it just that we never updated Legacy to use write commands?

Is it worth updating (in scope of another ticket) so that we can remove all this otherwise-dead code?

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/DeleteOpcodeOperation.cs]: We did update the Legacy to use write commands.

If write command are supported the DeleteOpcodeOperation will use a DeleteOpcodeOperationEmulator to use write commands while providing the DeleteOpcodeOperation interface and result type to the Legacy calling code.

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/DeleteOpcodeOperation.cs]: The Legacy code could instantiate a DeleteOpcodeOperationEmulator directly if we want to assume that write commands are supported.

Comment (jyemin) [reply src/MongoDB.Driver.Core/Core/Operations/DeleteOpcodeOperation.cs]: There is still the check on `writeConcern.IsAcknowledged` to consider. As I understand it, this class will continue to use OP_DELETE for unacknowledged writes even when OP_MSG is available.  Or is there a separate indirection to turn it into a one-way OP_MSG?

For the (non-legacy) Driver module, what protocol is used to execute unacknowledged writes on server versions 2.6 to 3.4 (inclusive) and for 3.6 to 4.2 (inclusive)?  In particular, is it the same or different as what's done for Driver Legacy?

File: src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/BinaryEncoders/BinaryMessageEncoderFactory.cs (hunk 6 above)
Comment (MikalaiMazurenka) [reply src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/BinaryEncoders/BinaryMessageEncoderFactory.cs]: Done

File: src/MongoDB.Driver.Core/Core/Operations/AggregateOperation.cs (hunk 3 above)
Comment (MikalaiMazurenka) [reply src/MongoDB.Driver.Core/Core/Operations/AggregateOperation.cs:346]: Done

File: src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/BinaryEncoders/BinaryMessageEncoderFactory.cs (hunk 6 above)
Comment (MikalaiMazurenka) [reply src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/BinaryEncoders/BinaryMessageEncoderFactory.cs]: Done

File: src/MongoDB.Driver.Core/Core/Operations/DeleteOpcodeOperation.cs (hunk 4 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/DeleteOpcodeOperation.cs]: Good point. I totally missed the part about unacknowledged writes.

I'll research the other question about unacknowledged writes in the new API.

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/DeleteOpcodeOperation.cs]: When using the new API we have support for unacknowledged writes using write commands (no need for OP_DELETE or anything like that).

See this line:

https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs#L111

We only wait for and process a response if one is expected.

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/DeleteOpcodeOperation.cs]: Note: "CommandMessage" == OP_MSG

Comment (jyemin) [reply src/MongoDB.Driver.Core/Core/Operations/DeleteOpcodeOperation.cs]: OK.  Let's leave this alone at least for this ticket.  But let's open a new ticket to remove what should now be dead code if not for the Legacy implementation, which I think could be re-written to use the same write operation as Driver (with some result transformation)

Comment (MikalaiMazurenka) [reply src/MongoDB.Driver.Core/Core/Operations/DeleteOpcodeOperation.cs]: From 2.6 to 3.6 the logic is a bit different, but it also doesn't use OP_INSERT, OP_DELETE or OP_UPDATE or any of classes in question (https://github.com/mongodb/mongo-csharp-driver/blob/4f791f7184552d0f87b4bd0b6d062cdde37f7d24/src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingQueryMessageWireProtocol.cs#L143)
Starting from 3.6 (feature "CommandMessage": https://github.com/mongodb/mongo-csharp-driver/blob/4f791f7184552d0f87b4bd0b6d062cdde37f7d24/src/MongoDB.Driver.Core/Core/Misc/Feature.cs#L47) it works as Robert described.
Summing all up, am I assuming correctly that we stick to deprecation instead of deleting for this ticket and create a ticket to completely remove operation classes and update the code which is not in legacy to use `BulkMixedWriteOperation`?

File: src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs (hunk 1 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs]: Our C# code that says CreateIndexesCommand was first supported in server 3.0 might be wrong.

Server documentation says it is new in server version 2.6:

https://docs.mongodb.com/v2.6/reference/command/createIndexes/

File: src/MongoDB.Driver.Core/Core/Operations/DeleteOpcodeOperation.cs (hunk 4 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/DeleteOpcodeOperation.cs]: I expected all traces of OP_DELETE, OP_INSERT and OP_UPDATE to be removed. They are no longer needed.

Unfortunately we have to keep support for OP_QUERY because the initial isMaster command must be sent using OP_QUERY.

**Review by rstam — CHANGES_REQUESTED** [review]
See previous comments.

File: src/MongoDB.Driver.Core/Core/Operations/AggregateOperation.cs [hunk 7]
```diff
@@ -253,6 +253,7 @@ public bool RetryRequested
         /// <value>
         /// A value indicating whether the server should use a cursor to return the results.
         /// </value>
+        [Obsolete("UseCursor has been deprecated since server version 3.4.")]
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Operations/AggregateOperation.cs]: This is really unrelated to CSHARP-2032.

It should have been done as part of CSHARP-1903.

File: src/MongoDB.Driver.Core/Core/Misc/Feature.cs [hunk 8]
```diff
@@ -41,12 +41,12 @@ public class Feature
...
-        private static readonly Feature __changeStreamPostBatchResumeToken = new Feature("ChangeStreamPostBatchResumeToken", new SemanticVersion(4, 0 ,7));
+        private static readonly Feature __changeStreamPostBatchResumeToken = new Feature("ChangeStreamPostBatchResumeToken", new SemanticVersion(4, 0, 7));
         private static readonly Feature __clientSideEncryption = new Feature("ClientSideEncryption", new SemanticVersion(4, 1, 9));
         private static readonly CollationFeature __collation = new CollationFeature("Collation", new SemanticVersion(3, 3, 11));
         private static readonly Feature __commandMessage = new Feature("CommandMessage", new SemanticVersion(3, 6, 0));
         private static readonly CommandsThatWriteAcceptWriteConcernFeature __commandsThatWriteAcceptWriteConcern = new CommandsThatWriteAcceptWriteConcernFeature("CommandsThatWriteAcceptWriteConcern", new SemanticVersion(3, 3, 11));
-        private static readonly Feature __createIndexesCommand = new Feature("CreateIndexesCommand", new SemanticVersion(3, 0, 0));
+        private static readonly Feature __createIndexesCommand = new Feature("CreateIndexesCommand", new SemanticVersion(2, 6, 0));
```
Comment (MikalaiMazurenka) [inline src/MongoDB.Driver.Core/Core/Misc/Feature.cs:49]: Had incorrect `firstSupportedVersion` (https://docs.mongodb.com/v2.6/reference/command/createIndexes/)

File: src/MongoDB.Driver.Core/Core/Operations/AggregateOperation.cs (hunk 3 above)
Comment (MikalaiMazurenka) [reply src/MongoDB.Driver.Core/Core/Operations/AggregateOperation.cs:346]: Done in scope of https://github.com/rstam/mongo-csharp-driver/pull/142

**Comment by MikalaiMazurenka** [comment]
Ready for review!

File: src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs [hunk 9]
```diff
@@ -124,7 +122,11 @@ public BsonDocument Execute(IWriteBinding binding, CancellationToken cancellatio
             using (var channel = channelSource.GetChannel(cancellationToken))
             using (var channelBinding = new ChannelReadWriteBinding(channelSource.Server, channel, binding.Session.Fork()))
             {
-                var operation = CreateOperation(channel.ConnectionDescription.ServerVersion);
+                var operation = new CreateIndexesUsingCommandOperation(_collectionNamespace, _requests, _messageEncoderSettings)
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs]: I would keep the CreateOperation helper method so we can continue to share code between the Sync and Async Execute methods.

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/AggregateOperationTests.cs [hunk 10]
```diff
@@ -343,26 +341,21 @@ public void UseCursor_get_and_set_should_work()
             {
                 BatchSize = batchSize
             };
-            var serverVersion = Feature.AggregateCursorResult.SupportedOrNotSupportedVersion(useServerVersionSupportingAggregateCursorResult);
-
+            var serverVersion = Feature.AggregateCursorResult.FirstSupportedVersion;
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Operations/AggregateOperationTests.cs:342]: Remove `useServerVersionSupportingAggregateCursorResult` test method parameter lines 337-338

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexesOperationTests.cs [hunk 11]
```diff
@@ -67,44 +67,6 @@ public void constructor_should_throw_when_messageEncoderSettings_is_null()
             argumentNullException.ParamName.Should().Be("messageEncoderSettings");
         }
 
-        [Theory]
-        [ParameterAttributeData]
-        public void CreateOperation_should_return_expected_result(
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexesOperationTests.cs:72]: Another reason to keep the CreateOperation helper method in CreateIndexesOperation, so that we can test operation creation.

I would reinstate this test, but removing the `isCommandSupported` test method parameter.

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexesOperationTests.cs [hunk 12]
```diff
@@ -274,7 +236,6 @@ public void constructor_should_throw_when_messageEncoderSettings_is_null()
         public void Execute_should_send_session_id_when_supported(
             [Values(false, true)] bool async)
         {
-            RequireServer.Check().Supports(Feature.CreateIndexesCommand);
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexesOperationTests.cs:277]: Keep the RequireServer.Check() part:

    RequireServer.Check();

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexesUsingCommandOperationTests.cs [hunk 13]
```diff
@@ -187,7 +187,6 @@ public void CreateCommand_should_return_expected_result_when_MaxTime_is_Set(long
             [Values(false, true)]
             bool async)
         {
-            RequireServer.Check().Supports(Feature.CreateIndexesCommand);
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexesUsingCommandOperationTests.cs:190]: Keep the RequireServer.Check() part

Same for all other methods were this happened.

File: tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs [hunk 14]
```diff
@@ -2414,18 +2349,9 @@ public void TestInsertBatchMultipleBatchesWriteConcernEnabledContinueOnErrorFals
...
-                Assert.Equal(1, results.Length);
-                Assert.Equal(true, results[0].HasLastErrorMessage);
-            }
-            else
-            {
-                Assert.Equal(2, results.Length);
-                Assert.Equal(false, results[0].HasLastErrorMessage);
-                Assert.Equal(true, results[1].HasLastErrorMessage);
-            }
+            // it the opcode was emulated there will just be one synthesized result
```
Comment (rstam) [inline tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs]: Let's fix minor typo (that was already there...):

    // because the opcode was emulated there will just be one synthesized result

File: tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs [hunk 15]
```diff
@@ -2469,19 +2395,9 @@ public void TestInsertBatchMultipleBatchesWriteConcernEnabledContinueOnErrorTrue
...
-                Assert.Equal(true, results[0].HasLastErrorMessage);
-            }
-            else
-            {
-                Assert.Equal(3, results.Length);
-                Assert.Equal(false, results[0].HasLastErrorMessage);
-                Assert.Equal(true, results[1].HasLastErrorMessage);
-                Assert.Equal(false, results[2].HasLastErrorMessage);
-            }
+            // it the opcode was emulated there will just be one synthesized result
```
Comment (rstam) [inline tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs]: Same as line 2352

File: tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs [hunk 16]
```diff
@@ -3491,22 +3401,19 @@ public void TestValidateWithFull()
         [Fact]
         public void TestValidateWithMaxTime()
         {
-            if (_primary.Supports(FeatureId.MaxTime) && _primary.InstanceType != MongoServerInstanceType.ShardRouter)
```
Comment (rstam) [inline tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs:3494]: Keep the second part of the test (the shard router part)

File: src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs (hunk 9 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs]: And so that we can test creating the operation (see comment in test file).

**Review by rstam — CHANGES_REQUESTED** [review]
Looks good!

Only some minor comments.

File: src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs (hunk 9 above)
Comment (MikalaiMazurenka) [reply src/MongoDB.Driver.Core/Core/Operations/CreateIndexesOperation.cs]: Done.

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/AggregateOperationTests.cs (hunk 10 above)
Comment (MikalaiMazurenka) [reply tests/MongoDB.Driver.Core.Tests/Core/Operations/AggregateOperationTests.cs:342]: Done.

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexesOperationTests.cs (hunk 11 above)
Comment (MikalaiMazurenka) [reply tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexesOperationTests.cs:72]: Done.

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexesOperationTests.cs (hunk 12 above)
Comment (MikalaiMazurenka) [reply tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexesOperationTests.cs:277]: Done for all removed `RequireServer.Check();` occurrences.

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexesUsingCommandOperationTests.cs (hunk 13 above)
Comment (MikalaiMazurenka) [reply tests/MongoDB.Driver.Core.Tests/Core/Operations/CreateIndexesUsingCommandOperationTests.cs:190]: Done for all removed `RequireServer.Check();` occurrences.

**Review by MikalaiMazurenka — COMMENTED** [review]
Ready for review!

File: tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs (hunk 14 above)
Comment (MikalaiMazurenka) [reply tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs]: Done.

File: tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs (hunk 15 above)
Comment (MikalaiMazurenka) [reply tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs]: Done.

File: tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs (hunk 16 above)
Comment (MikalaiMazurenka) [reply tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs:3494]: Done.

**Review by rstam — APPROVED** [review]
LGTM

**Review by DmitryLukyanov — APPROVED** [review]
